### ISSUE TITLE
glmer.nb()  still doubtful 

### ISSUE THREAD

**Issue by mmaechler** [lme4/lme4#319]
After all the closed issues about `glmer.nb()` problems which I have fixed now, notably the #287 one where we now do get correct estimates of fixed and random effects,...

there's still a basic misfunctioning,  notably exposed by the simple artificial example in `example(glmer.nb)`  -  get the latest git version of today!

The simulated example actually has "zero" random effects (not in the model, but the simulated data),
and hence `MASS :: glm.nb(.)`  works quite fine, finding "good" fixed effect estimates (and neg.bin dispersion parameter `theta`).
For some reason, in that example, the fixed effects estimated from our  `glmer.nb()` are ridiculously wrong.

**Comment by bbolker** [comment]
yikes.  at least we're digging into it.  thanks for checking.

At some point it will/would be interesting to do a systematic comparison of lognormal-Poisson vs NB, along the lines of the reference below (warning: I haven't looked at it closely, I don't actually know what I think of it!)

@article{harrison_comparison_2015,
	title = {A comparison of observation-level random effect and {Beta}-{Binomial} models for modelling overdispersion in {Binomial} data in ecology \&amp; evolution},
	volume = {3},
	issn = {2167-8359},
	url = {https://peerj.com/articles/1114},
	doi = {10.7717/peerj.1114},
	language = {en},
	urldate = {2015-07-21},
	journal = {PeerJ},
	author = {Harrison, Xavier A.},
	month = jul,
	year = {2015},
	pages = {e1114}
}

**Comment by bbolker** [comment]
I can't reproduce.  With most recent git version and running the code below I get `[1] "Mean relative difference: 4.158119e-06"` for the comparison of fixed effects.  (on both 32-bit Ubuntu 12.04 r-devel and 64-bit MacOS R 3.1.2)

```
set.seed(101)
dd <- expand.grid(f1 = factor(1:3),
                  f2 = LETTERS[1:2], g=1:9, rep=1:15,
                  KEEP.OUT.ATTRS=FALSE)
summary(mu <- 5*(-4 + with(dd, as.integer(f1) + 4*as.numeric(f2))))
dd$y <- rnbinom(nrow(dd), mu = mu, size = 0.5)

require("MASS")
m.glm <- glm.nb(y ~ f1*f2, data=dd, trace=TRUE)
summary(m.glm)
library(lme4)
m.nb <- glmer.nb(y ~ f1*f2 + (1|g), data=dd, verbose=TRUE)
all.equal(fixef(m.nb),coef(m.glm))
```

**Comment by bbolker** [comment]
I've pushed this example as a test in `inst/tests/test-glmernb.R`; we'll see how it does on Travis ...

**Comment by bbolker** [comment]
@mmaechler : bump?  have you looked into this any further?

**Comment by mmaechler** [comment]
Well, your new `test-glmernb.R`  does indeed break   'R CMD check lme4' for me...  I do get the nonsensically small values I had mentioned earlier.
Both platforms I have tried are  Fedora Linux (one, FC 21, my newish laptop;  the other my desktop or one of our compute servers, both runing FC 20)... so in some sense the version of gcc is "somewhat" old,  but not really.   E.g., on the notebook it is  gcc-Version 4.9.2 20150212 (Red Hat 4.9.2-6) (GCC) .

The fact that it passes on Travis means it does work on even older gcc (I think)... hmm, I have no clue.
Martin

**Comment by bbolker** [comment]
Maybe post `verbose` results so we can see where it's departing?  

Linux results: (R Under development (unstable) (2015-07-29 r68752); Platform: i686-pc-linux-gnu (32-bit); Running under: Ubuntu precise (12.04.5 LTS); gcc (Ubuntu/Linaro 4.6.4-1ubuntu1~12.04) 4.6.4)

```
theta.ml: iter 0 'theta = 0.470115'
theta.ml: iter1 theta =0.482275
theta.ml: iter2 theta =0.482681
theta.ml: iter3 theta =0.482681
th := est_theta(glmer(..)) = 0.4826813 --> dev.= -2*logLik(.) = 5969.284 
 1: th=   0.2377340525, dev= 6140.60807137
 2: th=   0.9800077574, dev= 6192.26658769
 3: th=  0.09906383297, dev= 6721.48252158
 4: th=   0.4547081036, dev= 5970.38221170
 5: th=   0.4605939030, dev= 5969.91396596
 6: th=   0.4826727026, dev= 5969.28348056
 7: th=   0.4798351618, dev= 5969.26927968
 8: th=   0.4797599933, dev= 5969.26927267
 9: th=   0.4797700781, dev= 5969.26927250
10: th=   0.4797780796, dev= 5969.26927261
11: th=   0.4797700781, dev= 5969.26927250
```

MacOS results are very nearly (but not quite) identical: (freshly installed lme4 via `devtools::install_github()`; R version 3.2.1 (2015-06-18); Platform: x86_64-apple-darwin13.4.0 (64-bit); Running under: OS X 10.9.5 (Mavericks); gcc info: Configured with: --prefix=/Applications/Xcode/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1; Apple LLVM version 5.1 (clang-503.0.40) (based on LLVM 3.4svn); Target: x86_64-apple-darwin13.4.0; Thread model: posix)

```
theta.ml: iter 0 'theta = 0.470115'
theta.ml: iter1 theta =0.482275
theta.ml: iter2 theta =0.482681
theta.ml: iter3 theta =0.482681
th := est_theta(glmer(..)) = 0.4826813 --> dev.= -2*logLik(.) = 5969.284 
 1: th=   0.2377340479, dev= 6140.60808030
 2: th=   0.9800077384, dev= 6192.26657488
 3: th=  0.09906383105, dev= 6721.48253779
 4: th=   0.4547081059, dev= 5970.38221149
 5: th=   0.4605939030, dev= 5969.91396596
 6: th=   0.4826727026, dev= 5969.28348056
 7: th=   0.4798351618, dev= 5969.26927968
 8: th=   0.4797599933, dev= 5969.26927267
 9: th=   0.4797700781, dev= 5969.26927250
10: th=   0.4797780796, dev= 5969.26927261
11: th=   0.4797700781, dev= 5969.26927250
> all.equal(fixef(m.nb),coef(m.glm))
[1] "Mean relative difference: 4.158119e-06"
```

**Comment by aosmith16** [comment]
Thanks for the hard work on `glmer.nb` @mmaechler.  I can reproduce the tiny fixed effects for this example in `glmer.nb`on my machine (Platform: x86_64-w64-mingw32/x64 (64-bit)) if I use the default optimizers.  

However, I get what look like reasonable results for the fixed effects if I switch to just using a single optimizer instead of `c("bobyqa", "Nelder_Mead")`.  I tested it with `bobyqa`, `Nelder_Mead`, and `nloptwrap` and each seems to work fine.

**Comment by mmaechler** [comment]
My `verbose=TRUE` output looks basically identical to yours.
The bug must be somewhere deep down: that "deepcopy" or other such magic low-level (Rccp) update magic does not work in my case...  but in your.

**Comment by bbolker** [comment]
The fact that `verbose=TRUE` output is nearly identical actually seems **extremely** encouraging to me, because it implies that something is going wrong right near the end (e.g. in the last stage of copying/reporting), not somewhere in the middle. If I had a system where this failed I would try to work on it, but (conveniently) I feel like I'm off the hook if I can't reproduce it ...  Do you know if the problem occurs for all test cases or only sometimes?  (If "only sometimes" I could try some brute-force/stress testing to see if I can find a case where it breaks on my platforms ...)

**Comment by mmaechler** [comment]
yeah... you are right.. in this sense it's good news.   Note that yesterday, I took my previous generation notebook with  ubuntu 14.04 LTS  ... all updated to the -- for 14.04 LTS -- latest gcc (4.8.4 IIRC), and I *did* see the same problem.   As ubuntu 14.04 LTS is available as docker image... everyone should - in principle - be able to reproduce it then..

**Comment by bbolker** [comment]
A little more progress: running on Ubuntu 14.04 64-bit with clang 3.5, I can reproduce.   Full test case posted [here](https://github.com/lme4/lme4/blob/99754ce2540228b61f96b4dbe5be98a30571ffcc/misc/issues/nbinom_refit.R),  but this is the key bit (the rest is just data generation equivalent to the `?glmer.nb` example:
```
m1 <- glmer(y ~ f1*f2 + (1|g), data=dd, family=poisson)
th <- lme4:::est_theta(m1,limit=20,eps=1e-4,trace=FALSE)
m2 <- update(m1,family=negative.binomial(theta=th))
m3 <- update(m1,family=negative.binomial(theta=0.4826813))
all.equal(m1@beta,(m1B <- refit(m1))@beta)
all.equal(m2@beta,(m2B <- refit(m2))@beta)  ## fails
all.equal(m3@beta,(m3B <- refit(m3))@beta)
```
So `refit` performs differently if we set `negative.binomial(theta=th)` vs `negative.binomial(theta=0.4826813)` ... I guess something's getting evaluated in the wrong environment??

**Comment by stevencarlislewalker** [comment]
I just pulled and ran all the code in this issue, and got no errors with this setup:

```{r}
> sessionInfo()
R version 3.2.0 (2015-04-16)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: OS X 10.9.5 (Mavericks)

locale:
[1] en_CA.UTF-8/en_CA.UTF-8/en_CA.UTF-8/C/en_CA.UTF-8/en_CA.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] lme4_1.1-9   Matrix_1.2-0 MASS_7.3-40

loaded via a namespace (and not attached):
[1] compiler_3.2.0  minqa_1.2.4     Rcpp_0.11.6     splines_3.2.0  
[5] nlme_3.1-120    grid_3.2.0      nloptr_1.0.4    lattice_0.20-31
```

**Comment by brendan-R** [comment]
I can reproduce with Ubuntu 15.04 (gcc 4.9.2-10ubuntu13). lme4 downloaded from master just now (at 5e7e6b4f21aa213c001a17dfe65112cd2de0f1f2).

```R
fitting glmer models : ...........................................................
glmer.nb : th := est_theta(glmer(..)) = 0.4826813 --> dev.= -2*logLik(.) = 5969.284 
 1: th=   0.2377340479, dev= 6140.60808031, beta[1]=    0.00000028
 2: th=   0.9800077384, dev= 6192.26657845, beta[1]=   -0.00327524
 3: th=  0.09906383105, dev= 6721.48253781, beta[1]=    0.00000551
 4: th=   0.4547081041, dev= 5970.38221166, beta[1]=   -0.00000117
 5: th=   0.4605939017, dev= 5969.91396605, beta[1]=    0.00000095
 6: th=   0.4826727032, dev= 5969.28348057, beta[1]=    0.00000095
 7: th=   0.4798351616, dev= 5969.26927968, beta[1]=   -0.00000161
 8: th=   0.4797599931, dev= 5969.26927267, beta[1]=   -0.00000237
 9: th=   0.4797700712, dev= 5969.26927250, beta[1]=   -0.00000237
10: th=   0.4797780727, dev= 5969.26927261, beta[1]=   -0.00000237
11: th=   0.4797700712, dev= 5969.26927250, beta[1]=   -0.00000237
.1
```

```R
> sessionInfo()
R version 3.2.1 (2015-06-18)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 15.04

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=C               LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] ggplot2_1.0.1   boot_1.3-17     MASS_7.3-43     lme4_1.1-9      Matrix_1.2-2   
[6] testthat_0.10.0

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.0         xml2_0.1.1          magrittr_1.5        roxygen2_4.1.1     
 [5] devtools_1.8.0.9000 splines_3.2.1       munsell_0.4.2       colorspace_1.2-6   
 [9] lattice_0.20-33     minqa_1.2.4         plyr_1.8.3          stringr_1.0.0      
[13] tools_3.2.1         parallel_3.2.1      grid_3.2.1          nlme_3.1-121       
[17] gtable_0.1.2        git2r_0.10.1.9000   rversions_1.0.2     digest_0.6.8       
[21] RcppEigen_0.3.2.5.0 crayon_1.3.1        reshape2_1.4.1      nloptr_1.0.4       
[25] codetools_0.2-11    curl_0.9.1          memoise_0.2.1       stringi_0.5-5      
[29] scales_0.2.5        proto_0.3-10    
```

I also got some other errors that I didn't understand, but as I saw the green Travis badge on the README, I thought I should mention them. [Here's](https://gist.github.com/brendan-R/edb71c9e9ba1214a269f) the error blurb from `devtools::test()`.

**Comment by bbolker** [comment]
More information on this: with very slightly varying initial `theta` values I can provoke it on different platforms, even those I thought were working.

For example (this is a non-working excerpt from https://github.com/lme4/lme4/blob/6951ac5a99f0e8d4df2c6db75960f6db02b749ca/misc/issues/nbinom_refit.R ):

```
mfun <- function(x) isTRUE(all.equal(x,m.symth@beta))
rmat <- matrix(ncol=10,nrow=length(fixef(m.symth)))
for (i in 1:10) {
    cat("*")
    m.numthXX <- update(m.base,
                        family=negative.binomial(theta=0.4826813+i*1e-10))
    tt <- system.time(rmat[,i] <- refit(m.numthXX)@beta)
    cat(" ",mfun(rmat[,i]),tt["elapsed"],"\n")
}
```

this gives different TRUE/FALSE patterns on different machines, e.g.

```
##  Ubuntu 14.04 64-bit:
##  [1]  TRUE  TRUE  TRUE  TRUE  TRUE FALSE FALSE  TRUE FALSE FALSE
##  Ubuntu 12.04 32-bit:
##  [1]  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE
```

At least this makes it slightly easier to debug because one can in
principle run side-by-side shells on the same machine and compare ...

**Comment by pitakakariki** [comment]
I've had a quick look and I noticed a couple of things.

a) The examples that fail take a bit longer to finish. This suggests that they're spinning through more iterations.

b) The difference between `theta=th` and `theta=0.4826813` is driven by numerical differences in the argument rather than how it's specified, at least in the examples I've tried. e.g. if I use `th0 <- 0.4826813` and then `theta=th` I get the same result as `th=0.4826813`.

So I suspect this is numerical stuff rather than copying or environments?

**Comment by pitakakariki** [comment]
I've been looking at the code in https://github.com/lme4/lme4/issues/319#issuecomment-127802556 and I might have found something:

    head(m2@resp$offset)
    # [1] 1.650082 2.417236 2.661554 3.162502 3.314590 3.563568

    head(m3@resp$offset)
    # [1] 0 0 0 0 0 0

i.e. the failing case (`m2`) has offset values where it shouldn't (see `m3`). Looks like that's messing up the objective function when `refit` calls Nelder Mead via `optwrap`. 

There's an open bug https://github.com/lme4/lme4/issues/334 which has to do with offsets appearing where they shouldn't, and I think this is related.

As noted in the offset bug report, the offset values are the predictions:

    all.equal(m2@resp$offset, unname(predict(m2)))
    # [1] TRUE

**Comment by bbolker** [comment]
examples above do work, but there is indeed some kind of general problem with the combination of `refit` and offsets: this is simple enough to provoke it:

```
library(lme4)
 m5 <- glmer(round(Reaction) ~ Days + (1|Subject),
                    data = sleepstudy, family=poisson,
                    offset=rep(0,nrow(sleepstudy)))
m5
refit(m5,debug=TRUE)
```

**Comment by bbolker** [comment]
fixed by  3c0fa76
